docs/design/proposals/destructive-operations/destructive-operations.md

```diff
@@ -0,0 +1,202 @@
+
+Another consideration of destructive operations is that some resources should continue to live on after the ACK CR has been deleted. Similar to the CDK[`RemovalPolicy` type](https://docs.aws.amazon.com/cdk/api/v2//docs/aws-cdk-lib.RemovalPolicy.html#members) and the native [K8s PV reclaim policy](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaiming), ACK should support the ability to leave the AWS resource alive while still deleting the custom resource object. Users typically use these removal policy options when working with stateful data, such as S3, EBS and RDS resources, in cases where they would like to migrate the management of the resources from one system to another without having to recreate the data.
+
+This design document proposes different mechanisms to address each of the following concerns:
+
+* Protection against deletion
+* Control over AWS removal policy
+
+## Delete Protection
+
+### Requirements
+
+1. Multiple levels of granularity over resource protection (namespace-wide and resource-wide)
+2. Protection should be defined per-service (if namespace-wide)
+3. Controller must strictly check for existence of resource protection before any deletion action
+4. The controller must not edit any of the resource protection marks
+
+### Investigation
+
+Kubernetes finalizers are the standard way to protect a resource from deletion. A finalizer placed on a resource prohibits it from being deleted out of etcd until it is removed - either through code or by manually editing the resource. ACK already places finalizers on every resource, however these are used to guard the custom resource object being deleted before the the underlying AWS resource. Finalizers, however, do not apply in a cascading effect to resources below them ie. a finalizer placed on a namespace will not protect the deletion of a custom resource within it. This violates requirement 1, as there is only resource-wide granularity over resource protection. 
+
+Finalizers do not fit with requirement 3. If the user places a finalizer on an ACK resource, then issues a deletion of that resource, the ACK controller will immediately delete the AWS resource and then remove its own finalizer, regardless of any other finalizer on the object. The current ACK controller does not look for other finalizers on the object before running its delete logic. Only the K8s control plane uses finalizers to protect it from deletion in the key-value store, it does not limit the interaction of custom controllers.
+
+There is on ongoing issue regarding similar functionality in the [Kubernetes Github repository](https://github.com/kubernetes/kubernetes/issues/10179).
```

Comment:
There is a reason that k/k issue 10179 is 7 years old and still not implemented. And that, IMHO, is because many developers realized that trying to keep users from fat-fingering things and accidentally deleting things is a completely futile effort. Users will always do stupid things. 99% of users will learn from their mistake. Interfaces and user experiences should not cater to babysitting and hand-holding imperative-style user interactions. Instead, interfaces and user experiences should focus on having no surprises and being hard to misuse.

In the case of ACK and the user experience of resource deletion, I posit that the behaviour of our interface has no surprises and is actually hard to misuse. It doesn't have surprises because our interface to delete a resource is to call `kubectl delete` on that resource. This is as non-surprising as it gets. It is hard to misuse because calling `kubectl delete` is an explicit action the user has to take and it's not a confusing verb or semantic. "delete" means to remove the resource.